### PR TITLE
[WIP] Replace docker.io as the default registry with registries[0]

### PR DESCRIPTION
**Pull request by rhatdan**
Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
This is an alternative to https://github.com/containers/image/pull/448

File: docker/reference/normalize.go [hunk 1]
```diff
@@ -29,11 +29,18 @@ type normalizedNamed interface {
...
+func ParseNormalizedNamed(sys *types.SystemContext, s string) (Named, error) {
 	if ok := anchoredIdentifierRegexp.MatchString(s); ok {
 		return nil, fmt.Errorf("invalid repository name (%s), cannot specify 64-byte hexadecimal strings", s)
 	}
-	domain, remainder := splitDockerDomain(s)
+	registries, err := GetRegistries(sys)
+	if err != nil {
+		return nil, err
+	}
+	if len(registries) == 0 {
```
Comment (TomSweeneyRedHat) [inline docker/reference/normalize.go]: Should we default to quay.io in this case?  Or fedora?

File: docker/reference/normalize.go [hunk 2]
```diff
@@ -29,11 +29,18 @@ type normalizedNamed interface {
...
+		return nil, errors.New("no registries specified")
```
Comment (TomSweeneyRedHat) [inline docker/reference/normalize.go]: "no registries specified in /etc/containers/registries.conf" perhaps?

**Comment by TomSweeneyRedHat** [comment]
This is more in line with what I was thinking you were going to do.  Looks like you've a whitespace error somewhere.  I'd go forward with this and kill the earlier one.

File: docker/reference/normalize.go (hunk 2 above)
Comment (rhatdan) [reply docker/reference/normalize.go]: The problem is I am not sure if that is the path.

File: docker/reference/normalize.go (hunk 1 above)
Comment (rhatdan) [reply docker/reference/normalize.go]: I really don't think the containers/image should have a default.  In the perfect world we would just allow a "" as the registry.

Comment (mtrmac) [reply docker/reference/normalize.go]: A minimal comment for now:

`c/i/docker/reference.Named`, `!IsNameOnly`, is a type representing a _specific_ image in a _specific_ registry.

Maybe there needs to be a type representing the “unqualified” inputs to search.

Maybe some uses of `reference.Named` should not use it at all, and accept more lax input. (In particular I was wrong about `containers-storage:`, it uses `reference.ParseNormalizedNamed` and does not allow hostname-less references.)

But, if hostname-qualified and hostname-less references should live side by side, they _need_ to be distinct types. The experience, _both_ with projectatomic/docker’s `reference`, and with podman’s `string` uses, is that mixing them together makes it all to easy to be careless about the semantics and results in surprising and unexpected corner case behaviors.

**Comment by rhatdan** [comment]
I think I prefer the original.  No registry at all.

**Pull request closed without merging**